### PR TITLE
[BLOG] Bril to RISC-V Lowering System

### LINKED ISSUE #407: Proposal: BRIL to RISC-V Lowering System
**What will you do?**

We will develop a system for lowering Bril programs to RISC-V assembly code, enabling the execution of Bril programs in a RISC-V environment. We will also focus on verifying the accuracy and correctness of this lowering system.

**How will you do it?**

To achieve this goal, we will do the following:

1. **Familiarization**: We will better acquaint ourselves with both the Bril language and the RISC-V architecture. One team member is actively working on implementing a RISC-V processor in Verilog for another course, and we feel as though we have a solid foundation in Bril. However, we will dive deeper into both before we begin implementation.

2. **Lowering Design**: We will design a system for lowering Bril instructions into RISC-V assembly code. Initially, our focus will be on creating a minimum viable implementation. In this phase, we will use a 1-to-N approach, where one Bril instruction corresponds to multiple RISC-V instructions. This basic version will serve as our foundation.

3. **Optimization (Stretch Goal)**: If time permits, we will explore an M-to-N approach. This strategy will involve optimizing the translation by recognizing dependencies between multiple Bril instructions and generating a more efficient sequence of RISC-V instructions. 

**How will you empirically measure success?**

We will evaluate the success of our project by assessing two key criteria:

1. **Semantic Equivalence**: We will verify the correctness of our project by ensuring that the translated RISC-V code is semantically equivalent to the source Bril code. We will utilize [Cranelift's](https://en.wikipedia.org/wiki/Cranelift) ISLE domain-specific language to specify our translation rules and use [Crocus](https://www.cs.cornell.edu/~avh/veri-isle-preprint.pdf) to verify the correctness of these rules. It's imperative that the RISC-V code generated by our system matches that produced by Cranelift. This will be tested using rigorous benchmarks.

2. **Performance Improvement**: We will assess the improvement in performance from interpreted Bril code to compiled RISC-V code in terms of execution time. We are currently researching tools to accurately measure wall clock time for comparing the two execution modes. 

**Team members:**
@20ashah

### PR DESCRIPTION
**Pull request by jdroob**
Closes #407

### CODE REVIEW

**Comment by jdroob** [comment]
*Update*

I just remembered we have an assembler available to us in ECE 4750 that works on a subset of RV32I. Out of my own curiosity, I'm going to look into this. Some tweaks need to be made to get the assembler to accept the RISC-V programs we're currently producing (e.g. some directives may need to be injected into our asm files, we need to use the 'x' register names instead of the special names such as 'sp').

File: content/blog/2023-12-11-lowering/index.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+Since we implemented trivial register allocation before function call lowering, we first tested our compiler thoroughly with [test cases](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend/test/trivial-reg-alloc) that encompassed every Bril instruction minus function calls / arguments. Once we verified that this worked, we moved on to testing with function calls once the calling conventions part was implemented. The main thing we tested was that we could execute this RISCV assembly and it was semantically equivalent to the Bril program.
+
+Just to highlight our results in some way, [here](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend/asm) are some examples that illustrate outputted RISCV assembly that we generated from Bril programs.
+
+---
+_John Rubio is a 2nd year MSCS student interested in hardware and compilers. In his free-time, he trains Brazilian Jiu-Jitsu._
+
+_Arjun Shah is a senior undergraduate majoring in CS. Arjun is interested in large-scale software systems. In his free time, Arjun trains competitive weight-lifting._
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: These can go in the designated spot for your bios in the front matter, as illustrated in this repo's README:
https://github.com/sampsyo/cs6120#cs-6120

File: content/blog/2023-11-28-mlir/index.md
Comment (sampsyo) [inline content/blog/2023-11-28-mlir/index.md:1]: Please delete this extra file, which belongs to a different PR.

File: content/blog/2023-12-11-lowering/index.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+[Codebase](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend)
+
+- [Bril Insn Classes + convert to RISCV functions](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend/BrilInsns)
+- [RISC Insn Classes](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend/RVIRInsns)
+- [Trivial Register Allocation Logic](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend/TrivialRegAlloc)
+- Calling Convention Logic
+  - [Prologue Inserter](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend/util/prologue.py)
+  - [Epilogue Inserter](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend/util/epilogue.py)
+  - [Lowering function call](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend/BrilInsns/BrilFunctionCallInsn.py)
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: Hmm… can you include these links more elegantly within the body of your writeup? It doesn't seem like people will be very interested in clicking on these links here at the beginning, before we have understood what you have done.

File: content/blog/2023-12-11-lowering/index.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+# The goal
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: It seems like this could merge with the overview sentence at the top? No need for a section heading here.

File: content/blog/2023-12-11-lowering/index.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+The goal for this project was to write the backend of a compiler that takes a bril program and lowers it to RISCV assembly - which ultimately
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: bril -> Bril (please find & replace throughout)

File: content/blog/2023-12-11-lowering/index.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+[extra]
+latex = true
+[[extra.authors]]
+name = "John Rubio"
+[[extra.authors]]
+name = "Arjun Shah"
++++
+
+# Summary
+@jdr299 (John Rubio) and I wrote a Python program that converts Bril to runnable RISCV assembly for our final project.
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: RISCV -> RISC-V (please replace throughout)

Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: You can just say "we". It's presumed that the text is speaking for both of you.

File: content/blog/2023-12-11-lowering/index.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+## Bril Instruction classes
+
+The first step in this project included writing classes for each Bril instruction. For code reuse, we organized these classes in a hierarchy that would 
+allow for similarities in instructions to be in common classes using inheritance. The main idea for this is if we needed to tweak these common aspects in 
+the instruction classes, the change would be in one place rather than having to keep track of all the different files to update. The other reason why we dealt 
+with objects rather than bril text or bril JSON, was because it is easier to keep track and update operands in an object format.
+
+The class hierarchy we came up with was the following:
+
+<img width="1689" alt="Screenshot 2023-12-11 at 6 46 26 PM" src="https://github.com/20ashah/cs6120/assets/33373825/5b165c72-39bf-44d0-93b5-2dc37a265bb9">
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: Can you please include all images with your post, here in this repo?

File: content/blog/2023-12-11-lowering/index.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+but our first goal was a working version. We wanted to implement this lowering as a 1-N approach, where each bril instruction would be mapped 
+to a certain number of RISCV instructions to get an abstract assembly. And then simply implementing calling conventions and trivial register allocation. 
+Each of these parts are described in detail below.
+
+## Bril Instruction classes
+
+The first step in this project included writing classes for each Bril instruction. For code reuse, we organized these classes in a hierarchy that would 
+allow for similarities in instructions to be in common classes using inheritance. The main idea for this is if we needed to tweak these common aspects in 
+the instruction classes, the change would be in one place rather than having to keep track of all the different files to update. The other reason why we dealt 
+with objects rather than bril text or bril JSON, was because it is easier to keep track and update operands in an object format.
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: Can you give a little more detail about what this inheritance hierarchy consisted of? You already have an exhaustive tree in your figure, of course, but it could be helpful to pick out one example and explain why introducing a superclass was helpful (vs. just a flat hierarchy where all instructions are equal) for a specific instance of code reuse.

File: content/blog/2023-12-11-lowering/index.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+An important note here is that in the first pass, we implemented this for every Bril instruction object except for function calls. The reason for this is that we wanted to save 
+the calling convention pass of lowering until the end, even after register allocation. We anticipated that this part would be the hardest, and so we wanted to get RISCV assembly 
+without function calls working first before we added that whole layer of complexity. This worked out quite well, since we could test the correctness of the assembly we were generating
+without function calls early on, without having implemented lowering for function calls.
+
+Below is a brief description of the conversions from Bril instructions to RISCV instructions. We filled out this table prior to actually coding these functions to make sure 
+that logically, our conversions created RISCV instructions that were semantically equivalent to the Bril. Note that this is abstract assembly, so no actual RISCV registers are used.
+
+
+| Bril                            | RISCV Abstract Asm                |
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: I think this table would be quite a bit easier to read if the code were contained in Markdown code blocks… like `` `this` `` or with `` ``` `` triple backticks for multi-line blocks.

File: content/blog/2023-12-11-lowering/index.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+| x: bool = or y z                | or x, y, z                        |
+| print x                         | nop                               |
+| jmp .label                      | jal x0, .label                    |
+| br cond .label1 .label2         | addi tmp, x0, 1 <br> beq cond, tmp, .label1 <br> jal x0, .label2 <br> .label1: <br> ... <br> jal x0 .exit <br> .label2: <br> ... <br> .exit: |
+| ret x                           | addi a0, x, 0 <br> jalr x0, x1, 0 |
+| ret                             | jalr x0, x1, 0                    |
+| x: int = id y                   | addi x, y, 0                      |
+ 
+                                  
+Note: An important note about the above chart that required some extra implementation had to deal with the case when the RISCV instructions needed to add in temps and labels
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: Probably no need for _both_ "Note:" and "an important note…". :smiley: In fact, you probably don't need either—you can just tell us the thing without telling us that it's a note!

File: content/blog/2023-12-11-lowering/index.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+| jmp .label                      | jal x0, .label                    |
+| br cond .label1 .label2         | addi tmp, x0, 1 <br> beq cond, tmp, .label1 <br> jal x0, .label2 <br> .label1: <br> ... <br> jal x0 .exit <br> .label2: <br> ... <br> .exit: |
+| ret x                           | addi a0, x, 0 <br> jalr x0, x1, 0 |
+| ret                             | jalr x0, x1, 0                    |
+| x: int = id y                   | addi x, y, 0                      |
+ 
+                                  
+Note: An important note about the above chart that required some extra implementation had to deal with the case when the RISCV instructions needed to add in temps and labels 
+to match the behavior of the Bril instructions. An edge case here is that these labels and temp variables need to be generated fresh each time for semantic equivalence, so 
+keeping track of this was a key part of the converter.
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: Can you point to one specific row of the table where this happens, so we can get a clearer picture of what you're referring to here?

File: content/blog/2023-12-11-lowering/index.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+implementation is not very efficient, since this requires loads and stores for every abstract assembly instruction, but it is the simplest implementation to create working assembly code.
+
+### Implementation
+
+The way this was implemented in our case was first creating a class that dealt with keeping track of stack offsets for each variable. What this class did was go through our list
+of RISCV instruction objects and for each one, pull out all of the abstract temps - variables that were not real RISCV registers and would need to be included in trivial register
+allocation. For each of these temps, this mapping class would assign it a unique offset in increments of 8 (starting from -8). 
+
+This was implemented as a dictionary in Python. So, if we come across a variable that is already in this dictionary, this means that we have seen it before and had already assigned it
+an offset, so we move on. Note that this happens on a per-function basis, so if we had two separate functions and they both contain a variable 'a', this variable is getting assigned
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: I'd use `` `a` `` instead of `'a'` because you are referring to a name from the code.

File: content/blog/2023-12-11-lowering/index.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+Now, we iterate through each RISCV instruction and get the list of abstract temps that it reads from and writes to. For each of the abstract temps that the instruction reads from, we use our mapper to look through the dictionary and find the stack offsets for each. Before this instruction, we add loads from these stack offsets into our RISCV registers. Then we perform the operation in the instruction, being sure to replace the abstract temps that are read from with the RISCV registers that we just loaded into. Now that the abstract temps in the instruction that we read from are taken care of, we move on to writes. We look at the abstract temp we write to, and replace that in the instruction with the next available RISCV register we haven't used for the reads yet. We then find using the mapper the stack location of this destination variable and we add an instruction after the operation that performs a store from the destination RISCV register into the appropriate stack offset the mapper gave us.
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: This may be a silly question, but if you're using trivial register allocation, why do you need this step (and, in fact, "abstract temps" at all)?

> Then we perform the operation in the instruction, being sure to replace the abstract temps that are read from with the RISCV registers that we just loaded into.

That is, the snippets of RISC-V code that are emitted could be _hard-coded_ to use `a0`, `a1`, and `a2`, since the operands always go in these locations when they are loaded from the stack. Then you wouldn't need to replace any abstract register names.

File: content/blog/2023-12-11-lowering/index.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+Doing this for each RISCV instruction gives us executable assembly that is complete without any abstract temps besides dealing with function calls/arguments and calling conventions. 
+
+## Calling Conventions
+
+By far, the biggest implementation obstacle was implementing the RISC-V calling conventions. We based our implementation off of the RISC-V calling conventions taught in Cornell's CS 3410: Computer System Organization and Programming. To oversimplify four lectures' worth of material, the essence of RISC-V calling conventions can be distilled into three main phases:
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: Maybe link to a webpage for 3410?

File: content/blog/2023-12-11-lowering/index.md [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+1. **Prologue (Function Entry Setup):** Before entering the function body, a set of preparatory actions, known as the prologue, take place. These include the creation of the stack frame. The size of the stack frame usually needs to be known ahead of time, considering the space needed to accommodate the return address, frame pointer, any overflow arguments, and local variables that must be pushed onto the stack. Additionally, all callee-save registers that will be utilized during the function execution should be pushed onto the stack at this stage.
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: This chunk of the blog post might be a little clearer if you phrased things in terms of what you _did_, as opposed to what (abstractly) needs to be done for any implementation of function calls. As it stands, it's hard to tell what you have actually implemented, versus what is still "to do." For example, you say:

> The size of the stack frame usually needs to be known ahead of time

Is this true in your compiler? Did you calculate the size of the stack frame?

File: content/blog/2023-12-11-lowering/index.md [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+2. **Function Call Instruction Preparation:** Prior to a function call instruction, certain steps must be taken to ensure the proper transfer of control and data. This involves placing function arguments in designated registers, following the argument-passing registers convention. For additional arguments or those that don't fit into the designated registers, space on the stack is allocated to hold these values. The function call instruction is then executed, initiating the transfer of control to the callee.
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: And here, can you be specific about whether you implemented spills to the stack from the registers for the calling convention?

File: content/blog/2023-12-11-lowering/index.md [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+
+3. **Epilogue (Clean-up):** The third key piece of this calling convention puzzle is the clean-up step, or the epilogue. After the instructions in the function body have finished executing, the epilogue is responsible for restoring the stack to its original state and releasing any resources allocated during the prologue. This includes popping the stack frame, restoring the values of callee-save registers, and ensuring a smooth return to the calling function.
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: Once again, it's not clear how helpful this general background about calling conventions is… it would be better to be specific about the case of your implementation, RISC-V, and Bril.

File: content/blog/2023-12-11-lowering/index.md [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
...wered Bril [programs](https://github.com/JohnDRubio/CS_6120_Advanced_Compilers/tree/main/rv32_backend/test) to RISC-V using our system. Next, we wrote C++ programs that were semantically equivalent to the same Bril programs and obtained the compiled RV32 code using [godbolt](https://godbolt.org/). Lastly, we ran both our RISC-V program and the godbolt RISC-V program on RISC-V interpreters and compared the end architectural state of each program. On the programs we tested, our RISC-V programs yielded identical architectural states to the godbolt programs, save quite a few more items on the stack in our case since we used trivial register allocation. Obviously, there are some issues with this approach - it does not provide a high degree of coverage and it does not provide any details on performance. We concede these two points. This approach does show that, at the very least, our lowering system can produce RISC-V programs that are semantically equivalent to non-trivial Bril programs. We would've liked to explore performance speedups in further detail but ultimately information on the performance of the Javascript runtime and that of compiled RISC-V programs is not difficult to find.
```
Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: What is called "godbolt" here is actually named "Compiler Explorer".

Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: Can you be specific about how many programs you tried, and how you selected them?

Comment (sampsyo) [inline content/blog/2023-12-11-lowering/index.md]: Javascript -> JavaScript

**Review by sampsyo — CHANGES_REQUESTED** [review]
Alright; cool that you got something minimal working here! As you note in the post, the main missing piece is a testing setup that can automatically compare the results of running the RISC-V code against some other implementation (e.g., the reference interpreter). But it seems like you got some of the basics in place, which is great.

File: content/blog/2023-11-28-mlir/index.md
Comment (jdroob) [reply content/blog/2023-11-28-mlir/index.md:1]: MLIR file has been removed.

File: content/blog/2023-12-11-lowering/index.md (hunk 1 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: Bios have been moved to the correct locations.

File: content/blog/2023-12-11-lowering/index.md (hunk 3 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: I completely agree - 'Goal' and 'Summary' sections merged.

File: content/blog/2023-12-11-lowering/index.md (hunk 17 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: Changed to Compiler Explorer.

File: content/blog/2023-12-11-lowering/index.md (hunk 6 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: Sorry about that! Moved to correct location :)

File: content/blog/2023-12-11-lowering/index.md (hunk 7 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: I've updated the figure to reflect the current class hierarchy and tried to provide more context. Happy to provide more if you feel it'd be helpful!

File: content/blog/2023-12-11-lowering/index.md (hunk 14 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: That makes sense - I've added discussion on what we actually did in our implementation in the Prologue, FunctionCall, and Epilogue sub-sections.

File: content/blog/2023-12-11-lowering/index.md (hunk 8 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: Agreed. Let me know if the current change works.

File: content/blog/2023-12-11-lowering/index.md (hunk 2 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: Agreed and (finally) moved links to more natural locations :)

File: content/blog/2023-12-11-lowering/index.md (hunk 10 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: I've updated the blog post and this excerpt is no longer part of the post. However, a similar point is made in the current version in the 'Challenges Along the Way' subsection (approximately line 101). Here, we provide a concrete example.

File: content/blog/2023-12-11-lowering/index.md (hunk 12 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: Embarrassingly enough, we simply didn't think of this...😅 This makes a ton of sense too since we know exactly which RISC-V instructions each Bril instruction maps to. This is definitely a good "work smarter not harder" moment for us to learn from lol.

File: content/blog/2023-12-11-lowering/index.md (hunk 17 above)
Comment (jdroob) [reply content/blog/2023-12-11-lowering/index.md]: We've added information describing how many programs were tested and why they were chosen in the 'Challenges Along the Way' section.

**Review by jdroob — COMMENTED** [review]
All requested changes should've been covered. Plus quite a bit of clean up :) Please let us know if there's anything else we should fix!

**Review by sampsyo — APPROVED** [review]
Awesome; nice work on the revisions!
